projectq/ops/_gates.py

```diff
@@ -338,3 +338,59 @@ def get_inverse(self):
 
 #: Shortcut (instance of) :class:`projectq.ops.BarrierGate`
 Barrier = BarrierGate()
+
+
+class FlipBits(SelfInverseGate):
+    """ Gate for transforming qubits in state |0> to any computational basis state """
+    def __init__(self, bits_to_flip):
+        """
+        Initialize FlipBits gate.
+
+        Example:
+            .. code-block:: python
+
+                qureg = eng.allocate_qureg(2)
+                FlipBits([0, 1]) | qureg
+
+        Note:
+            The amplitude of state k is final_state[k]. When the state k is
+            written in binary notation, then qureg[0] denotes the qubit
+            whose state corresponds to the least significant bit of k.
+
+        Args:
+            bits_to_flip(list[int]|list[bool]|str|int): int or array of 0/1, True/False, or
+                                    string of 0/1 of length up to len(qureg) identifying
+                                    the qubits to flip. In case of int, the bits to flip are
+                                    determined from the binary digits, with the least significant
+                                    bit corresponding to qureg[0]. If bits_to_flip is negative,
+                                    exactly all bits which would not be flipped for the input
+                                    -bits_to_flip are flipped.
+        """
+        if isinstance(bits_to_flip, int):
+            self.bits_to_flip = bits_to_flip
+        elif isinstance(bits_to_flip, str):
+            self.bits_to_flip = list([ c != "0" for c in bits_to_flip])
+        else:
+            self.bits_to_flip = list(bits_to_flip)
+
+    def __str__(self):
+        return "FlipBits("+str(self.bits_to_flip)+")"
+
+    def __or__(self, qubits):
+        for qureg in self.make_tuple_of_qureg(qubits):
```

Comment:
Is there a specific reason why FlipBits acts on each quantum register separately?

I would have either allowed a single quantum register only (with assertion), or concatenate the registers if several are provided. I think the current implementation may lead to confusion: `FlipBits(2) | (qubit1, qubit2)` vs `FlipBits(2) | (qubit1 + qubit2)`.

Reply:
I just mirrored the behavior of `MeasureGate`. I still would like to support all the qureg/qubit formats I test in `_simulator_test.py` and calling `make_tuple_of_qureg()` seems to be the easiest way of doing this (please correct me if I am wrong here), also this means that changes to that function will be automatically supporting.
The behavior for `FlipBits(2) | (qubit1, qubit2)` was indeed confusing and as this notation is (as far as I understand) meant to be used only for controlled gates, I now raise a ValueError if the input is a tuple of len > 1. If I try to execute `FlipBits(2) | (qubit1 + qubit2)`, I get `unsupported operand type(s) for +: 'Qubit' and 'Qubit'`, so I am not exactly sure what you had in mind here.
The tests should make it pretty clear which notations for the right hand side of `__or__()` are now supported. Let me know if you want anything further changed.

Reply:
Great, thanks!

I was assuming that qubit1 and qubit2 were allocated via `eng.allocate_qubit()`, which returns a quantum register of length 1 and so + would concatenate the two quantum registers, resulting a register of length 2.
The notation `gate | (qureg1, qureg2)` is by no means reserved for controlled gates only. For example, addition of two registers would be `Add | (qureg1, qureg2)`.

Reply:
Ok, I understand. Then do you prefer that I support `FlipBits(x) | (qureg1, qureg2)` and treat it in the same way as `FlipBits(x) | qureg1 + qureg2` or do you prefer the current implementation that raises a `ValueError`?

Reply:
Either is fine imo

Reply:
I like the current restriction to a single argument and otherwise a ValueError but would be fine with either.